login: Split alert into title/description

martinpitt opened the pull request:
This makes it conformant to the current PF design.

Bring the DOM structure closer to current PF5's alert, and
adjust/simplify CSS accordingly. Thanks Garrett for working this out!

Also tighten the error message checks in the tests, in particular for
check-ws-bastion.

Rework the "no-cockpit" message: Telling the user which OSes are
supported is nice at some level, but doesn't help operationally when
they want to connect to *that* server. Thus, suggest to install
cockpit-system, which is the recommended way out of this error.

----

Reproducer for showing the banner on login page (do this in a test VM):
```
printf '[Session]\nBanner = /tmp/banner\n' | sudo tee -a /etc/cockpit/cockpit.conf
printf 'Hello from my laptop\nsecond line\n' > /tmp/banner
sudo systemctl stop cockpit
# so that cockpit can read /tmp/banner
sudo setenforce 0
```

Banner:

![image](https://github.com/user-attachments/assets/a3d7c4d6-66c2-4248-8487-9aa8d6143ab6)

Login alert without message (title only):

![image](https://github.com/user-attachments/assets/eb86ecae-fde4-48f7-82d5-c80ced120d8e)

Login alert with message (unknown server, or wrong password):

![image](https://github.com/user-attachments/assets/1600aec0-2991-4b2b-91f2-0599d01a7c79)

Redesigned "no cockpit" error when in beiboot mode:

![Screen Shot 2024-10-24 at 08 50 13](https://github.com/user-attachments/assets/0b76e917-0b42-498c-bebe-d3abc2d09270)

martinpitt commented:
First version -- the spacing in the alert is all wrong. For an alert with a message, like trying to connect to host "nothing.local" (and user name a, empty password is ok) it has too much gap and margin (even though I already set them to 0):

![image](https://github.com/user-attachments/assets/936aac8c-3a7c-46b2-b637-8b13b5da246b)

And for a message without details, like trying to log in without an username, the message bit is visible as space. I think that's fixed easily enough with hiding it if it's empty, unless we can do that with CSS:

![image](https://github.com/user-attachments/assets/951355a2-cae9-4244-80e5-dfbf9b662430)

@garrett I'm happy to spend some more work on the CSS, unless you already have some hints. I also need to adjust quite a number of tests, so doing a first run to see which ones fail now.

garrett commented on pkg/static/login.html:
```diff
@@ -30,9 +30,12 @@ <h1 id="brand" class="hide-before"></h1>
-      <h2 id="login-error-message" class="pf-v5-c-alert__title">
-        <span class="noscript" translate="yes">Please enable JavaScript to use the Web Console.</span>
-      </h2>
+      <p id="login-error-title" class="pf-v5-c-alert__title"></p>
+      <div class="pf-v5-c-alert__description">
+        <p id="login-error-message">
+          <span class="noscript" translate="yes">Please enable JavaScript to use the Web Console.</span>
+        </p>
+      </div>
```
As for layout, I would expect a block with this structure

- alert
   - icon
   - title
   - description (optional)

Then a CSS grid like this:

```css
.pf-v5-c-alert {
  display: grid;
  grid-template: "icon content" / auto 1fr;
  gap: 0.5rem 1rem;
  align-items: center;
}

.pf-v5-c-alert__title,
.pf-v5-c-alert__description {
  grid-column: content;
}
```

It basically sets up a grid to have two columns, where the first is automatically sized and the other column is sized to the rest of the space.

Details: The icon goes into the first available grid cell, which is sized to auto and is called "icon" (which doesn't matter really, as we're not referring to it otherwise; it's implicitly placed — we could explicitly tell it to go there, but it's fine as-is as long as it's the first child of the grid anyway). Then the title and description are told to go into the "content" column, which is the remainder space. Since there are only two columns, the third child of the grid is on another row (and would go to the icon column, except we told it to go to the content column instead). The gap adds space between everything (both vertical and horizontal gaps as shorthand), and that vertical gap it makes the description optional, as it's unused if there's no second row with the description element.

You can see a full demo @ https://codepen.io/garrett/pen/YzmVmdK

(This uses the default PF5 HTML for the alert, not adjusted or anything, so it should theoretically be mostly a drop-in to our CSS.)

martinpitt replied:
It's not 100% perfect yet, but

> Garrett: it's mostly good enough, definitely better than what was there before
> we could also do that in a cosmetic follow-up

So let's get that in, and handle the rest in a follow-up if we care.

garrett reviewed (changes requested):
I'm surprised at PF's implementation. I explained why. But anyway, I made some simple grid CSS that would support an optional description (if the element is there) that should be mostly a drop-in replacement, and provided a link to the demo. (We might need to remove a little existing CSS to make it fit, but it's probably mostly fine?)

Hopefully this'll fix the spacing issues.

cockpituous commented on pkg/static/login.js:
```diff
@@ -543,14 +543,15 @@ function debug(...args) {
             /* OAuth failures are always fatal */
             if (oauth) {
-                fatal(msg);
+                fatal(title);
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -905,12 +907,12 @@ function debug(...args) {
 
         function key_down(e) {
-            login_failure(null, "conversation");
+            login_failure(null, null, "conversation");
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1009,13 +1011,13 @@ function debug(...args) {
                         fatal(format(_("The server refused to authenticate '$0' using password authentication, and no other supported authentication methods are available."), user));
                     } else if (xhr.statusText.indexOf("terminated") > -1) {
-                        login_failure(_("Authentication failed: Server closed connection"));
+                        login_failure(_("Authentication failed"), _("Server closed connection"));
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1009,13 +1011,13 @@ function debug(...args) {
                         host_failure(_("Unable to connect to that address"));
                     } else if (xhr.statusText.indexOf("unknown-hostkey") > -1) {
-                        host_failure(_("Refusing to connect. Hostkey is unknown"));
+                        host_failure(_("Refusing to connect"), _("Hostkey is unknown"));
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,18 +1028,22 @@ function debug(...args) {
             } else if (xhr.status == 403) {
-                login_failure(_(decodeURIComponent(xhr.statusText)) || _("Permission denied"));
+                const status = decodeURIComponent(xhr.statusText).trim();
+                login_failure(_("Permission denied"), status === "Permission denied" ? "" : status);
```
This added line is not executed by any test.

martinpitt commented:
Next round. I incorporated @garrett 's design/demo, and adjusted it a bit (actual PF colors and 2px instead of 3px top border width). The login failure alerts look quite fine now:

![image](https://github.com/user-attachments/assets/d78e3551-c831-482b-9645-8c4719b75541)

![image](https://github.com/user-attachments/assets/b5db7281-6ec7-45ec-bf84-8aa8509bbad4)

However, the icon is now centered with the description (or title) if it has multiple lines. Easily seen with setting 

```
printf '[Session]\nBanner = /tmp/banner\n' | sudo tee -a /etc/cockpit/cockpit.conf
printf 'Hello from my laptop\nsecond line\n' > /tmp/banner
sudo systemctl stop cockpit
```

On main:

![image](https://github.com/user-attachments/assets/bd3b8793-a428-4843-a81a-40b7e5f82425)

This PR:

![image](https://github.com/user-attachments/assets/f9125360-d6b0-43db-991d-58f684044929)

This also affects the changed host key alert in the same way. 

I'll look at that next. But I wanted to do a full round of tests in between, as the previous round still had a failure on arch, and I changed some DOM structure.

cockpituous commented on pkg/static/login.js:
```diff
@@ -543,15 +543,17 @@ function debug(...args) {
             /* OAuth failures are always fatal */
             if (oauth) {
-                fatal(msg);
+                fatal(title);
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -905,12 +909,12 @@ function debug(...args) {
 
         function key_down(e) {
-            login_failure(null, "conversation");
+            login_failure(null, null, "conversation");
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1009,13 +1013,13 @@ function debug(...args) {
                         fatal(format(_("The server refused to authenticate '$0' using password authentication, and no other supported authentication methods are available."), user));
                     } else if (xhr.statusText.indexOf("terminated") > -1) {
-                        login_failure(_("Authentication failed: Server closed connection"));
+                        login_failure(_("Authentication failed"), _("Server closed connection"));
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1009,13 +1013,13 @@ function debug(...args) {
                         host_failure(_("Unable to connect to that address"));
                     } else if (xhr.statusText.indexOf("unknown-hostkey") > -1) {
-                        host_failure(_("Refusing to connect. Hostkey is unknown"));
+                        host_failure(_("Refusing to connect"), _("Hostkey is unknown"));
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,18 +1030,22 @@ function debug(...args) {
             } else if (xhr.status == 403) {
-                login_failure(_(decodeURIComponent(xhr.statusText)) || _("Permission denied"));
+                const status = decodeURIComponent(xhr.statusText).trim();
+                login_failure(_("Permission denied"), status === "Permission denied" ? "" : status);
```
This added line is not executed by any test.

garrett commented:
> However, the icon is now centered with the description (or title) if it has multiple lines. Easily seen with setting

Whoops! My fault. I think the vertical spacing is incorrect too. But these things should be easy to fix.

Reference screenshot from https://www.patternfly.org/components/alert/html/#inline-variations (yeah, it's success state instead of error, but it's inline with a description):

![image](https://github.com/user-attachments/assets/8a4fa6a0-27c3-4a65-ba5c-823df56b0431)

And for the long title, from https://www.patternfly.org/components/alert/html/#variations:

![image](https://github.com/user-attachments/assets/5cbdbc37-0417-4370-bd88-732312f11e22)

(The alignment isn't right in PF5 either, but it's closer.)

garrett commented:
Changes that should fix the icon alignment:

```css
.pf-v5-c-alert {
  align-items: start;
}

.pf-v5-c-alert__icon {
  margin-top: calc((1cap - 1ex) / -2);
}
```

martinpitt commented on pkg/static/login.js:
```diff
@@ -1026,24 +1030,22 @@ function debug(...args) {
             } else if (xhr.status == 500 && xhr.statusText.indexOf("no-cockpit") > -1) {
-                // always show what's going on
-                let message = format(_("A compatible version of Cockpit is not installed on $0."), login_machine || "localhost");
-                // in beiboot mode we get some more info
-                const error = JSON.parse(xhr.responseText);
-                if (error.supported)
-                    message += " " + format(_("This is only supported for $0 on the target machine."), error.supported);
-                login_failure(message);
+                login_failure(
+                    _("Cockpit not installed on remote host"),
+                    format(_("Install at least the cockpit-system package on $0 to connect."), login_machine || "localhost")
```
@garrett: I haven't quite made up my mind between "tell the user which remote OSes are supported" and "tell the user how to fix it" (install cockpit-system). "you can remotely connect to Fedora 40, Debian testing, and RHEL 9.6" seems "nice" at some level, but not in this particular situation: you want to connect to the given machine, not a random one which happens to have a compatible OS.

So I think like this it's more helpful. WDYT?

garrett replied:
What about something like this?

## Dynamic remote session unavailable

This system does not support dynamic remote sessions. Install the `cockpit-system` package on the remote system to manage it via the web console.

martinpitt replied:
summary from chat: so far my favourites are:

 (1) drop the first sentence and only leave the second
 (2) s/dynamic/package-less/ , as that's more descriptive of what this is about

garrett replied:
Alternative attempt:

## Packageless session unavailable

Transient packageless connections currently require $0. Install the `cockpit-system` package on the remote system to manage it via the web console.

martinpitt replied:
Looks like this now:

![Screen Shot 2024-10-23 at 12 02 10](https://github.com/user-attachments/assets/68bb913f-de8c-48c6-859f-83e6cb1c5406)

garrett replied:
The "at least" part really isn't good. It sounds super awkward.

Packageless session unavailable

Transient packageless sessions require the same operating system and version, Fedora Linux 40, for compatibility reasons. Install the `cockpit-system` package (and optionally other `cockpit` packages) on `10.111.113.2` to enable web console access. 

or

Packageless session unavailable

Transient packageless sessions require the same operating system and version, for compatibility reasons: Fedora Linux 40. Install the `cockpit-system` package (and optionally other `cockpit` packages) on `10.111.113.2` to enable web console access. 

(BTW: It's Fedora Linux 40, not "fedora 40". Note the capitalization and "Linux" in the official name.)

Also: I wonder if it's too dense and covers different topics and if each sentence should be its own paragraph? Or if we should just keep the alerts headline plus body.

martinpitt replied:
OK, thanks! Changed to your second version:

![Screen Shot 2024-10-24 at 08 50 13](https://github.com/user-attachments/assets/0b76e917-0b42-498c-bebe-d3abc2d09270)

As for the "paragraph or not": This would currently require bigger changes. But I'm going to enable it for multiple OSes as the next step anyway, and then it'll need to become a list. I suggest to postpone the paragraph splitting until then, is that ok?

garrett replied:
Sure! This is already way better than what we have in main. The spacing and alignment is better than in main too :wink:, even if we agree it should be improved also. This is good enough to merge, IMHO, and we can do the rest in a follow-up.

martinpitt commented:
@garrett With your original formula, it looks oddly off:

![image](https://github.com/user-attachments/assets/03fbb1b7-78be-45af-a946-345a01115b66)

![image](https://github.com/user-attachments/assets/e9eb2ff0-3e28-438e-8591-e731d36a0a98)

Is your intent to align them at the bottom line, not at the top?

when I turn the -2 into 2 to move the icon *down* instead of *up* (which feels a bit nicer to me), it looks like this:

![image](https://github.com/user-attachments/assets/48bf7d4a-2d2b-4a32-8dce-3b4b0a39b3ea)

![image](https://github.com/user-attachments/assets/a4e7745c-2ebd-452a-bed9-d78f48a7f992)

garrett commented:
OK, that's weird. 

I guess something's different in the sandbox vs. the live environment. The gap is also wrong weirdly. I'll need to look more. 

The basic approach still looks fine, but the details need some adapting/tuning apparently. (The problem of in vitro vs. in vivo. My CSS was working fine in the "petri dish", but clearly not here in context.)

martinpitt commented:
@garrett I pushed your suggestion for now (plus test adjustments). The red login errors are straightforward to reproduce. See https://github.com/cockpit-project/cockpit/pull/21130#issuecomment-2426085108 if you also want to see the blue banner (do all this in a test VM preferably)

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,24 +1030,22 @@ function debug(...args) {
             } else if (xhr.status == 403) {
-                login_failure(_(decodeURIComponent(xhr.statusText)) || _("Permission denied"));
+                const status = decodeURIComponent(xhr.statusText).trim();
+                login_failure(_("Permission denied"), status === "Permission denied" ? "" : status);
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,24 +1030,22 @@ function debug(...args) {
-                login_failure(message);
+                login_failure(
+                    _("Cockpit not installed on remote host"),
+                    format(_("Install at least the cockpit-system package on $0 to connect."), login_machine || "localhost")
```
This added line is not executed by any test.

martinpitt commented:
@garrett I pushed what I have, and put current screenshots into the description. I resolved/hid most comments here to make it less noisy. I think the alignment is still not quite perfect, but at least not completely wrong. I'd also like to confirm/discuss the verbiage around "no-cockpit" [above](https://github.com/cockpit-project/cockpit/pull/21130#discussion_r1810365038). Thanks!

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,24 +1030,32 @@ function debug(...args) {
             } else if (xhr.status == 403) {
-                login_failure(_(decodeURIComponent(xhr.statusText)) || _("Permission denied"));
+                const status = decodeURIComponent(xhr.statusText).trim();
+                login_failure(_("Permission denied"), status === "Permission denied" ? "" : status);
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,24 +1030,32 @@ function debug(...args) {
+                    login_failure(
+                        _("Packageless session unavailable"),
+                        format(_("Transient packageless connections currently require $0."), error.supported) + " " +
+                        format(_("Install at least the cockpit-system package on $0 to manage it via the web console."), login_machine || "localhost")
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,24 +1030,32 @@ function debug(...args) {
+                        format(_("Install at least the cockpit-system package on $0 to manage it via the web console."), login_machine || "localhost")
+                    );
+                } else {
+                    login_failure(
+                        _("Cockpit not installed on remote host"),
+                        format(_("Install at least the cockpit-system package on $0 to manage it via the web console."), login_machine || "localhost")
```
These 3 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,24 +1030,30 @@ function debug(...args) {
             } else if (xhr.status == 403) {
-                login_failure(_(decodeURIComponent(xhr.statusText)) || _("Permission denied"));
+                const status = decodeURIComponent(xhr.statusText).trim();
+                login_failure(_("Permission denied"), status === "Permission denied" ? "" : status);
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1026,24 +1030,30 @@ function debug(...args) {
-                // in beiboot mode we get some more info
+                let message = format(
+                    _("Install the cockpit-system package (and optionally other cockpit packages) on $0 to enable web console access."),
+                    login_machine || "localhost");
```
This added line is not executed by any test.

garrett reviewed (approved):
Thanks!!!

I think this is good to merge. It's definitely better than what's in main, and we do want to revisit it.

We should open an issue to address the remaining concerns, so we don't forget: The styling needs improvement, and we want to relax the requirement some, both of which means another iteration at least.

martinpitt commented:
Thanks @garrett for your help and the iterations! I track the remaining bits in https://issues.redhat.com/browse/COCKPIT-1178

The code review should hopefully be straightforward.
